### PR TITLE
Using custom HtmlEncoder in unit tests - Part 1

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #2085 
There are changes to rendering tests that are not included in this PR. Will send that out as part 2.
@Eilon @dougbu

### CODE REVIEW

**Comment by Eilon** [comment]
:shipit:

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperFormExtensionsTest.cs [hunk 1]
```diff
@@ -276,7 +276,7 @@ public static TheoryData<object> RouteValuesDataSet
         public void BeginFormWithNoParameters_CallsHtmlGeneratorWithExpectedValues()
         {
             // Arrange
-            var tagBuilder = new TagBuilder(tagName: "form", htmlEncoder: new HtmlEncoder());
+            var tagBuilder = new TagBuilder(tagName: "form", htmlEncoder: new NullTestEncoder());
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperFormExtensionsTest.cs]: does anything demonstrate `@Html.BeginForm()` correctly encodes values passed as arguments?

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/TagBuilderTest.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@ public class TagBuilderTest
         public void MergeAttribute_IgnoresCase(bool replaceExisting, string expectedKey, string expectedValue)
         {
             // Arrange
-            var tagBuilder = new TagBuilder("p", new HtmlEncoder());
+            var tagBuilder = new TagBuilder("p", new NullTestEncoder());
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/TagBuilderTest.cs]: `TagBuilder` deals directly with encoding.  is that tested?

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/OptionTagHelperTest.cs [hunk 3]
```diff
@@ -401,7 +401,14 @@ public static IEnumerable<object[]> DoesNotUseViewContextDataSet
...
                     var tagHelperContent = new DefaultTagHelperContent();
-                    tagHelperContent.SetContent(originalContent);
+                    if (string.IsNullOrEmpty(originalContent))
+                    {
+                        tagHelperContent.SetContent(originalContent);
+                    }
+                    else
+                    {
+                        tagHelperContent.SetContent("HtmlEncode[[" + originalContent + "]]");
+                    }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/OptionTagHelperTest.cs]: why is this doing a fake encoding at this level?  what fails if `originalContent` is assigned without this?

**Comment by dougbu** [comment]
:watch: for better encoding coverage in the test classes that are using just `NullTestEncoder`.

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperFormExtensionsTest.cs (hunk 1 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperFormExtensionsTest.cs]: It is tested here https://github.com/aspnet/Mvc/pull/2420/files#diff-8cd78f293ecf9bb7f3a377fc18b05d39R162

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/TagBuilderTest.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/TagBuilderTest.cs]: Added a test.

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.TagHelpers/OptionTagHelper.cs [hunk 4]
```diff
@@ -78,6 +78,10 @@ public class OptionTagHelper : TagHelper
                     {
                         selected = selectedValues.Contains(Value);
                     }
+                    else if (output.IsContentModified)
+                    {
+                        selected = encodedValues.Contains(output.Content.GetContent());
+                    }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/OptionTagHelper.cs]: sorry, this is about where the :dancers: danced

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/TagBuilderTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/TagBuilderTest.cs]: thx

**Comment by dougbu** [comment]
the following classes do not test using `CommonTestEncoder` at all.  please fill these gaps.
- `AnchorTagHelperTest`
- `FormTagHelperTest`
- `TagHelperOutputExtensionsTest`
- `ValidationMessageTagHelperTest`

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperFormExtensionsTest.cs (hunk 1 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperFormExtensionsTest.cs]: thx

**Comment by dougbu** [comment]
:watch:

**Comment by ajaybhargavb** [comment]
@dougbu Not sure what scenario actually uses encoding in these helpers. I couldn't find any part of the code that uses with encoding directly. Could you provide some scenarios? 

> the following classes do not test using CommonTestEncoder at all. please fill these gaps.
>
>AnchorTagHelperTest
>FormTagHelperTest
>TagHelperOutputExtensionsTest
>ValidationMessageTagHelperTest

**Comment by dougbu** [comment]
@ajaybhargavb the named tag helpers all get `TagBuilder` instances from `DefaultHtmlGenerator`.  they then at least call `MergeAttributes.  so the basic thing is to ensure extra encoding does not occur.  for example, use the `CommonTestEncoder` instead in tests that just check `output.Attributes` (and see nothing is encoded).  or, in tests that look at the final `string`, should see HTML encoding of the attribute values.

for `TagHelperOutputExtensionsTest` I recommend creating the `TagBuilder` instances with `CommonTestEncoder` as well.  won't get used but that is still testing the scenario.

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**